File: src/main/java/io/github/intellij/dlanguage/codeinsight/render/BasicRenderer.java
package io.github.intellij.dlanguage.codeinsight.render;

import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementPresentation;
import io.github.intellij.dlanguage.codeinsight.dcd.DCDCompletionClient;
import io.github.intellij.dlanguage.codeinsight.dcd.completions.DCDCompletionKind;
import io.github.intellij.dlanguage.codeinsight.dcd.completions.DCDModel;
import io.github.intellij.dlanguage.icons.DlangIcons;
import org.jetbrains.annotations.NotNull;

public class BasicRenderer extends LookupElement {
    DCDModel model;

    public BasicRenderer(DCDModel model) {
        this.model = model;
    }

    @NotNull
    @Override
    public String getLookupString() {
        if (model.kind == DCDCompletionKind.Function)
            return model.value + "()";
        else
            return model.value;
    }

    @Override
    public void renderElement(LookupElementPresentation presentation) {
        presentation.setIcon(DlangIcons.FILE);
        presentation.setItemText(model.value);
        presentation.setItemTextBold(true);
        // if typeText is not null that means it is user code
        // else it is language specific code
        if (model.typeText != null)
            presentation.setTypeText(model.typeText);
        else
            presentation.setItemTextBold(false);

        switch (model.kind) {
            case ClassName:
                presentation.setIcon(DlangIcons.NODE_CLASS);
                break;
            case InterfaceName:
                presentation.setIcon(DlangIcons.NODE_INTERFACE);
                break;
            case StructName:
                presentation.setIcon(DlangIcons.NODE_STRUCT);
                break;
            case UnionName:
                presentation.setIcon(DlangIcons.NODE_UNION);
                break;
            case VariableName:
                presentation.setIcon(DlangIcons.NODE_FIELD);
                break;
            case MemberVariableName:
                presentation.setIcon(DlangIcons.NODE_FIELD);
                break;
            case Keyword:
                presentation.setIcon(null);
                break;
            case Function:
                presentation.setIcon(DlangIcons.NODE_FUNCTION);
                break;
            case EnumName:
                presentation.setIcon(DlangIcons.NODE_ENUM);
                break;
            case EnumMember:
                presentation.setIcon(DlangIcons.NODE_FIELD);
                break;
            case PackageName:
                presentation.setIcon(null);
                break;
            case ModuleName:
                presentation.setIcon(DlangIcons.MODULE);
                break;
            case Array:
                break;
            case AssociativeArray:
                break;
            case AliasName:
                presentation.setIcon(DlangIcons.NODE_ALIAS);
                break;
            case TemplateName:
                break;
            case MixinTemplate:
                presentation.setIcon(DlangIcons.NODE_MIXIN);
                break;
        }
    }
}

File: src/main/kotlin/io/github/intellij/dlanguage/codeinsight/render/BasicRenderer.kt
package io.github.intellij.dlanguage.codeinsight.render

import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementPresentation
import io.github.intellij.dlanguage.codeinsight.dcd.DCDCompletionClient
import io.github.intellij.dlanguage.codeinsight.dcd.completions.DCDCompletionKind
import io.github.intellij.dlanguage.codeinsight.dcd.completions.DCDModel
import io.github.intellij.dlanguage.icons.DlangIcons

class BasicRenderer(internal var model: DCDModel) : LookupElement()
{

    override fun getLookupString(): String
    {
        return if (model.kind === DCDCompletionKind.Function)
            model.value + "()"
        else
            model.value
    }

    override fun renderElement(presentation: LookupElementPresentation)
    {
        presentation.icon = DlangIcons.FILE
        presentation.itemText = model.value
        presentation.isItemTextBold = true
        // if typeText is not null that means it is user code
        // else it is language specific code
        if (model.typeText != null)
            presentation.typeText = model.typeText
        else
            presentation.isItemTextBold = false

        when (model.kind)
        {
            DCDCompletionKind.ClassName -> presentation.icon = DlangIcons.NODE_CLASS
            DCDCompletionKind.InterfaceName -> presentation.icon = DlangIcons.NODE_INTERFACE
            DCDCompletionKind.StructName -> presentation.icon = DlangIcons.NODE_STRUCT
            DCDCompletionKind.UnionName -> presentation.icon = DlangIcons.NODE_UNION
            DCDCompletionKind.VariableName -> presentation.icon = DlangIcons.NODE_FIELD
            DCDCompletionKind.MemberVariableName -> presentation.icon = DlangIcons.NODE_FIELD
            DCDCompletionKind.Keyword -> presentation.icon = null
            DCDCompletionKind.Function -> presentation.icon = DlangIcons.NODE_FUNCTION
            DCDCompletionKind.EnumName -> presentation.icon = DlangIcons.NODE_ENUM
            DCDCompletionKind.EnumMember -> presentation.icon = DlangIcons.NODE_FIELD
            DCDCompletionKind.PackageName -> presentation.icon = null
            DCDCompletionKind.ModuleName -> presentation.icon = DlangIcons.MODULE
            DCDCompletionKind.Array ->
            {
            }
            DCDCompletionKind.AssociativeArray ->
            {
            }
            DCDCompletionKind.AliasName -> presentation.icon = DlangIcons.NODE_ALIAS
            DCDCompletionKind.TemplateName ->
            {
            }
            DCDCompletionKind.MixinTemplate -> presentation.icon = DlangIcons.NODE_MIXIN
        }
    }
}
